core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java

```diff
@@ -0,0 +1,112 @@
+package jenkins.slaves;
+
+import hudson.AbortException;
+import hudson.Extension;
+import hudson.model.Computer;
+import hudson.remoting.Channel;
+import hudson.remoting.ChannelBuilder;
+import hudson.slaves.SlaveComputer;
+import jenkins.AgentProtocol;
+import jenkins.model.Jenkins;
+import jenkins.security.ChannelConfigurator;
+import org.jenkinsci.remoting.engine.JnlpServer3Handshake;
+import org.jenkinsci.remoting.nio.NioChannelHub;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.net.Socket;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Master-side implementation for JNLP3-connect protocol.
+ *
+ * <p>@see {@link org.jenkinsci.remoting.engine.JnlpProtocol3} for more details.
+ *
+ * @author Akshay Dayal
+ * @since 1.XXX
+ */
+@Extension
```

Comment:
I'm going to spend some time dusting out the stress-test suites that we used before and run them for JNLP3. Will probably have to modify them a bit to have significant "work" being done for each build instead of just "sleep 5". Will let you know when I'm done.